### DISCUSSION TITLE
Custom API (golang)

### DISCUSSION THREAD

**Discussion by m-2k** [pocketbase/pocketbase#4020] [Q&A] [answered]
1. is the solution correct from the points of view of the Pocketbase concept?
2. can I use url paths starting with `/api/*`?
3. Is it possible to update the JSON in a field directly through the DB driver without reading the record? If so, the performance will be much higher, but does it make much sense? JSON field with `{ [user_id: string] => timestamp: string }`
4. Is it necessary to validate `requestData.PostIds` data before passing it into `FindRecordsByIds/2` (^[0-9a-z]{15}$ or existing PB validator)?

Thx!

```go
e.Router.AddRoute(echo.Route{
	Method: http.MethodPatch,
	Path:   "/custom",
	Handler: func(c echo.Context) error {
		type viewedField map[string]string
		const fieldNameViewed = "viewed"
		const collectionName = "messages"
		userId := apis.RequestInfo(c).AuthRecord.Id
		
		if err := validation.ValidateStruct(&requestData,
			validation.Field(&requestData.PostIds, validation.Required,
				validation.Each(validation.Match(regexp.MustCompile("^[0-9a-z]{15}$"))))); err != nil {
					return apis.NewBadRequestError("Validation error", err)}
		now := types.DateTime.String(types.NowDateTime())

		requestData := struct {
			PostIds []string `form:"postIds" json:"postIds"`
		}{}
		if err := c.Bind(&requestData); err != nil {
			return apis.NewBadRequestError("Failed to load the submitted data due to invalid formatting.", err)
		}

		dao := app.Dao().WithoutHooks()
		if err := dao.RunInTransaction(func(txDao *daos.Dao) error {
			records, err := txDao.FindRecordsByIds(collectionName, requestData.PostIds)
			if err != nil {
				return err
			}

			for _, record := range records {
				var viewed viewedField
				raw, ok := record.Get(fieldNameViewed).(types.JsonRaw)
				if !ok {
					return fmt.Errorf("Field %s not found in record of collection %s", fieldNameViewed, collectionName)
				}
				if err := json.Unmarshal(raw, &viewed); err != nil {
					return err
				}
				if viewed == nil {
					viewed = make(viewedField)
				}
				viewed[userId] = now
				record.Set(fieldNameViewed, viewed)

				if err := txDao.SaveRecord(record); err != nil {
					return err
				}
			}
			return nil
		}); err != nil {
			return apis.NewBadRequestError("Failed to execute query", nil)
		}

		return c.NoContent(http.StatusOK)
	},
	Middlewares: []echo.MiddlewareFunc{
		apis.RequireAdminOrRecordAuth(),
	},
})

return nil
})
```

```js
<button
on:click={async () => {
console.log(
	await pocketbase.send("/custom", {
		method: "PATCH",
		body: { postsId: ["s3n5c6fariay9ou"] },
	})
)
}}>Test</button>
```

**Comment by ganigeorgiev** [comment, marked as answer]
1. I'm not sure that I understand the purpose of the above action, nor I have the full picture of the use case to provide suggestions.

2. Yes. The recommendation usually if you want your routes to be under the `/api/` path is to add also the app name to prevent eventual collisions with future system routes (eg. `/api/yourApp/*`).

3. It depends on the type of the JSON change. For executing raw db queries you can check the [Executing queries](https://pocketbase.io/docs/go-database/#executing-queries) docs. For the available SQLite JSON functions you can explore https://www.sqlite.org/json1.html.

4. They should be safe to be assigned directly as arguments because the ids are added as named parameters to the query. If you really need validation, then I'd suggest only a `validation.Length(0. 100)` or similar to restrict the submitted string size, but it is not required. Also note that `FindRecordsByIds()` can succeed without error and still return empty slice. `sql.ErrNoRows` is usually returned only for single record or exec statement queries.

**Reply by ganigeorgiev** [reply]
_Side-note: you can also unmarshal directly a `json` record field in your struct using the [`record.UnmarshalJSONField(field, result)`](https://pkg.go.dev/github.com/pocketbase/pocketbase@v0.20.1/models#Record.UnmarshalJSONField) helper._

**Reply by m-2k** [reply]
> I'm not sure that I understand the purpose of the above action, nor I have the full picture of the use case to provide suggestions.

I'm making a chat for two (as a special case) or more people (groups) and I need to provide users with information about how opponents have read messages. Since messages can be read in random order (leaving holes of unread messages in the chat), I decided that I didn't just need to keep the last timestamp for each user within the chat, but store read information within each message. so I have a table with messages and a column `viewed` that stores json `{ userId => timestamp}`
From the client (js-sdk) I just want to add a read timestamp to this message field, but the timestamp must be generated on the server to avoid problems

**Reply by ganigeorgiev** [reply]
It would really depend who can have access to the `viewed` data and the overall desired data flow.

I have a similar "read/unread comment" scenario in Presentator (my other project) and I'm using a separate "notifications" collection with 2 `relation` fields for the "comment" and "user" (+2 additional flags for the read status and whether the message was "processed", aka. whether an email notification was sent for it as we don't send emails for already read comments). The records in this collection are created for each related user with an after hook when a new comment is added.

**Reply by m-2k** [reply]
Messages can only be accessed by those who have been added to the `chat` table field `members: []rel->users`.
I also place more emphasis on more universal architecture through the interaction of more than two members into chat.
Thank you for your answers, timely as always and with additional useful information.
